Test involving multiprocessing crashes when pytest 5.1.2 and xdist are used

This simple test `testCrashXdist1` crashs pytest when xdist is used. It seems to start when pytest==5.1.2 is used. When using pytest==5.1.1, pytest does not crash and ValueError is raised correctly. My OS is linux.

The most likely cause is this change although I'm not sure: https://github.com/pytest-dev/pytest/issues/5786


```python
def badFunc():
    raise ValueError()


def testCrashXdist1():
    import multiprocessing

    with multiprocessing.Pool(1) as pool:
        assert pool.apply(badFunc)


def testCrashXdist2():
    assert badFunc()
```

```
$ py.test testcrash.py -n1 -k testCrashXdist1
============================================================================================================ test session starts ============================================================================================================
platform linux -- Python 3.6.5, pytest-5.1.2, py-1.8.0, pluggy-0.13.0
rootdir: /tmp/crash
plugins: forked-1.1.1, xdist-1.29.0
gw0 [1]
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/main.py", line 235, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/remote.py", line 71, in pytest_runtestloop
INTERNALERROR>     self.run_one_test(torun)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/remote.py", line 85, in run_one_test
INTERNALERROR>     self.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/runner.py", line 72, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/runner.py", line 87, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/runner.py", line 171, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/remote.py", line 109, in pytest_runtest_logreport
INTERNALERROR>     config=self.config, report=report
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/reports.py", line 328, in pytest_report_to_serializable
INTERNALERROR>     data = report._to_json()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/reports.py", line 164, in _to_json
INTERNALERROR>     return _report_to_json(self)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/reports.py", line 394, in _report_to_json
INTERNALERROR>     d["longrepr"] = serialize_longrepr(report)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/reports.py", line 381, in serialize_longrepr
INTERNALERROR>     serialize_repr_crash(repr_crash),
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/reports.py", line 367, in serialize_repr_crash
INTERNALERROR>     return reprcrash.__dict__.copy()
INTERNALERROR> AttributeError: 'NoneType' object has no attribute '__dict__'
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/main.py", line 191, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/_pytest/main.py", line 235, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 92, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/manager.py", line 86, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/dsession.py", line 115, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/dsession.py", line 138, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/tmp/testpytestvenv/lib/python3.6/site-packages/xdist/dsession.py", line 180, in worker_workerfinished
INTERNALERROR>     assert not crashitem, (crashitem, node)
INTERNALERROR> AssertionError: ('testcrash.py::testCrashXdist1', <WorkerController gw0>)
INTERNALERROR> assert not 'testcrash.py::testCrashXdist1'

```

```
$ pip list
Package            Version
------------------ -------
apipkg             1.5
atomicwrites       1.3.0
attrs              19.3.0
execnet            1.7.1
importlib-metadata 0.23
more-itertools     7.2.0
packaging          19.2
pip                19.3
pluggy             0.13.0
py                 1.8.0
pyparsing          2.4.2
pytest             5.1.2
pytest-forked      1.1.1
pytest-xdist       1.29.0
setuptools         41.4.0
six                1.12.0
wcwidth            0.1.7
wheel              0.33.6
zipp               0.6.0
```

--- a/changelog/5971.bugfix.rst
+++ b/changelog/5971.bugfix.rst
@@ -1,0 +1,2 @@
+Fix a ``pytest-xdist`` crash when dealing with exceptions raised in subprocesses created by the
+``multiprocessing`` module.

--- a/src/_pytest/reports.py
+++ b/src/_pytest/reports.py
@@ -374,8 +374,11 @@ def _report_to_json(report):
         ]
         return result
 
-    def serialize_repr_crash(reprcrash):
-        return reprcrash.__dict__.copy()
+    def serialize_repr_crash(reprcrash: Optional[ReprFileLocation]):
+        if reprcrash is not None:
+            return reprcrash.__dict__.copy()
+        else:
+            return None
 
     def serialize_longrepr(rep):
         result = {
@@ -455,8 +458,11 @@ def _report_kwargs_from_json(reportdict):
         ]
         return ReprTraceback(**repr_traceback_dict)
 
-    def deserialize_repr_crash(repr_crash_dict):
-        return ReprFileLocation(**repr_crash_dict)
+    def deserialize_repr_crash(repr_crash_dict: Optional[dict]):
+        if repr_crash_dict is not None:
+            return ReprFileLocation(**repr_crash_dict)
+        else:
+            return None
 
     if (
         reportdict["longrepr"]

--- a/testing/test_reports.py
+++ b/testing/test_reports.py
@@ -305,11 +305,73 @@ class TestReportSerialization:
 
         data = report._to_json()
         loaded_report = report_class._from_json(data)
+
+        assert loaded_report.failed
         check_longrepr(loaded_report.longrepr)
 
         # make sure we don't blow up on ``toterminal`` call; we don't test the actual output because it is very
         # brittle and hard to maintain, but we can assume it is correct because ``toterminal`` is already tested
         # elsewhere and we do check the contents of the longrepr object after loading it.
+        loaded_report.longrepr.toterminal(tw_mock)
+
+    def test_chained_exceptions_no_reprcrash(
+        self, testdir, tw_mock,
+    ):
+        """Regression test for tracebacks without a reprcrash (#5971)
+
+        This happens notably on exceptions raised by multiprocess.pool: the exception transfer
+        from subprocess to main process creates an artificial exception which, ExceptionInfo
+        can't obtain the ReprFileLocation from.
+        """
+        testdir.makepyfile(
+            """
+            # equivalent of multiprocessing.pool.RemoteTraceback
+            class RemoteTraceback(Exception):
+                def __init__(self, tb):
+                    self.tb = tb
+                def __str__(self):
+                    return self.tb
+
+            def test_a():
+                try:
+                    raise ValueError('value error')
+                except ValueError as e:
+                    # equivalent to how multiprocessing.pool.rebuild_exc does it
+                    e.__cause__ = RemoteTraceback('runtime error')
+                    raise e
+        """
+        )
+        reprec = testdir.inline_run()
+
+        reports = reprec.getreports("pytest_runtest_logreport")
+
+        def check_longrepr(longrepr):
+            assert isinstance(longrepr, ExceptionChainRepr)
+            assert len(longrepr.chain) == 2
+            entry1, entry2 = longrepr.chain
+            tb1, fileloc1, desc1 = entry1
+            tb2, fileloc2, desc2 = entry2
+
+            assert "RemoteTraceback: runtime error" in str(tb1)
+            assert "ValueError('value error')" in str(tb2)
+
+            assert fileloc1 is None
+            assert fileloc2.message == "ValueError: value error"
+
+        # 3 reports: setup/call/teardown: get the call report
+        assert len(reports) == 3
+        report = reports[1]
+
+        assert report.failed
+        check_longrepr(report.longrepr)
+
+        data = report._to_json()
+        loaded_report = TestReport._from_json(data)
+
+        assert loaded_report.failed
+        check_longrepr(loaded_report.longrepr)
+
+        # for same reasons as previous test, ensure we don't blow up here
         loaded_report.longrepr.toterminal(tw_mock)
 
 

--- a/testing/test_reports.py
+++ b/testing/test_reports.py
@@ -320,25 +320,19 @@ class TestReportSerialization:
         """Regression test for tracebacks without a reprcrash (#5971)
 
         This happens notably on exceptions raised by multiprocess.pool: the exception transfer
-        from subprocess to main process creates an artificial exception which, ExceptionInfo
+        from subprocess to main process creates an artificial exception, which ExceptionInfo
         can't obtain the ReprFileLocation from.
         """
         testdir.makepyfile(
             """
-            # equivalent of multiprocessing.pool.RemoteTraceback
-            class RemoteTraceback(Exception):
-                def __init__(self, tb):
-                    self.tb = tb
-                def __str__(self):
-                    return self.tb
+            from concurrent.futures import ProcessPoolExecutor
+
+            def func():
+                raise ValueError('value error')
 
             def test_a():
-                try:
-                    raise ValueError('value error')
-                except ValueError as e:
-                    # equivalent to how multiprocessing.pool.rebuild_exc does it
-                    e.__cause__ = RemoteTraceback('runtime error')
-                    raise e
+                with ProcessPoolExecutor() as p:
+                    p.submit(func).result()
         """
         )
         reprec = testdir.inline_run()
@@ -352,8 +346,8 @@ class TestReportSerialization:
             tb1, fileloc1, desc1 = entry1
             tb2, fileloc2, desc2 = entry2
 
-            assert "RemoteTraceback: runtime error" in str(tb1)
-            assert "ValueError('value error')" in str(tb2)
+            assert "RemoteTraceback" in str(tb1)
+            assert "ValueError: value error" in str(tb2)
 
             assert fileloc1 is None
             assert fileloc2.message == "ValueError: value error"

